Enable Amendments from config file or static data

seelabs opened the pull request:
* The rippled.cfg file has a new section called "amendments"

* Each line in this section contains two white-space separated items
** The first item is the id of the amendment (a 256-bit hash)
** The second item is the friendly name

* Replaces config section name macros with variables

* Make addKnown arguments safer

* Added lock to addKnown

@scott @tomswirly

vinniefalco commented:
Commit messages should not drill down to the nuts and bolts of the language, i.e.:
"Changed function signature of addKnown from two char*'s  to a hash and a string, makes it less likely to call incorrectly."
Consider instead
"Make addKnown arguments safer"

seelabs commented:
Got it. Will change.

scottschurr commented on src/ripple/core/ConfigSections.h:
```diff
@@ -30,48 +30,48 @@ struct ConfigSection
-#define SECTION_WEBSOCKET_PING_FREQ     "websocket_ping_frequency"
-#define SECTION_VALIDATORS              "validators"
-#define SECTION_VALIDATORS_SITE         "validators_site"
+static std::string const SECTION_ACCOUNT_PROBE_MAX       ("account_probe_max");
```
Hey, Scott, here's a suggestion that may significantly expand the scope of your change.  All caps is completely appropriate for #defines.  Not so great for everything else.  So I think the names of these static const strings should be lower case.  That means you'll need to fix up all their references, of course.

My initial take, since I'm not familiar with this code, is you could make a `Section` namespace inside the `ripple` namespace and drop all of the names in there.  Then they'd be referenced like this: `Section::databasePath`.  Vinnie or David may suggest a better approach.  Or there may be a pre-existing class or namespace they should be inside.

vinniefalco replied:
I think its wasteful to create aliases for what is essentially a string. They should just be hard coded strings in the program. This way you can easily find them, without getting bounced to a header file. We have a problem with too much proliferation of symbols. You have to ask "whats the benefit versus the cost of adding another symbol". In the case of these config sections, it creates a centralized "header hotspot" which becomes a _cross cutting concern_:
http://en.wikipedia.org/wiki/Cross-cutting_concern
Adding new sections in this style becomes bureaucratic and cumbersome. There's also no real benefit to changing them to be string constants, and there could be a downside (Visual Studio might get link errors).

seelabs replied:
There are 114 lines containing "SECTION_". It doesn't seem like too much work. I'll hold off on this if others want to weigh in. I agree it makes the code better.

seelabs replied:
vinnie: Did I misunderstand your comment in the code (// VFALCO TODO Rename and replace these macros with variables.)? I took that to mean to do something like I did. So change them back to defines?

vinniefalco replied:
@seelabs I wrote that comment when I first started looking at the code, before I knew a lot. You should remove the comment instead.

miguelportilla replied:
The aliases would be good if we were localizing, but we are not.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -67,8 +80,9 @@ class AmendmentTableImpl : public AmendmentTable
 
-    AmendmentState* addKnown (const char* amendmentID, const char* friendlyName,
-        bool veto) override;
+    AmendmentState* addKnown (uint256 const& amendmentID,
```
I'm guessing that no one derives from the `AmendmentTableImpl` and that is wasn't really designed to be further derived from (I could be wrong).  If I'm right, the `override` on these methods is good, but would be even better as `final`.  As an alternative you could just mark the class as `final`.  I depends on how easy you want to make it for someone to choose to derive from `AmendmentTableImpl`.

seelabs replied:
Agree, will mark class as final. It lives in a cpp file. If someone wants to inherit from it later, it's self contained in the cpp, so it'll be easy to see the consequences of removing the "final".

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -22,10 +22,23 @@
+   server. The first element of the pair is the id, the second element is
+   the "friendly name".
+*/
+std::set<std::pair<std::string, std::string>> PreEnabledAmendments;
```
Is this an empty `set` for future use?  As far as I can tell it's only referenced by a `for` loop later on which (it looks like for now) will execute no iterations.  Is that your intention?

seelabs replied:
One of the requirements of RIPD-746 was "Read enabled features from static data embedded in the application". This collection supports that. The idea is people can add amendments here by changing the ::ctor from empty to one with an initializer list of id,name pairs.

rec replied:
This first sentence (which needs a . at the end) is a little tantalizing, particularly the quotes - is it compiled in or just "compiled in"?  and what is "it"?  ;-)

What about something like:

    PreEnabledAmendments is a static list of amendments that are enabled at build time.

rec replied:
First, we use exactly the same naming convention for static or global variables as any other - that is camelCase.   CapCamelCase is reserved for ClassNames.  (I think this is unfortunate - most other shops use ALL_CAPS for constants - but it is what it is).

I think using `std::pair<std::string, std::string>` is a poor choice to represent the id and "friendly name" - which I think we call "human-readable name".

It's far too easy write `am.first` instead of `am.second` somewhere down the line and a reviewer might well miss it.  We have seen exactly such problems (using tuples, actually) in rippled before.

With C++11 and uniform initialization, any old struct is just as good as a `std::pair`.  Consider something like:

    struct AmendmentName 
    {
        std::string id;
        std::string humanName;
    };

rec replied:
Also, you probably want a typedef for this set - you use the same type again later.

seelabs replied:
The reason preEnabledAmendments is a pair of string is because there is no way to tell if an invalid hex string was fed to the uint64(string) constructor. So if I change preEnabledAmendments to take a pair of <uint256,string> and initialize it with {uint256("123XXX"), "invalid_first"}, then the program will not spot the error. As written, it will throw on this error. The use-case for preEnabledAmendments is human-entered hashes, so throwing on an easily made typo is important.

I agree with your point about pairs of string,string easy to get wrong. This was the reason I changed the sig of addKnown. I could change the implementation of base_uint to try to handle this, but I'm *really* hesitant to much around with such a basic type. I think the string,string pair is actually less error prone for this use case. I think <string,string> pair is better for this case.

seelabs replied:
@rec Changing from a pair<string,string> to a struct (instead of a pair<uint256,string>) addresses this. I can check for a valid hex string in the ::ctor of this class. Will change.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +114,60 @@ class AmendmentTableImpl : public AmendmentTable
+                else
+                    friendlyName = curTok;
+            }
+
```
Interesting.  I'm about to pick a nit, since it looks like this code should work just fine.  The inner (`curTok`) `for` loop can handle an arbitrary number of arguments, but always expects 2.  Because this code is organized as a `for` loop you have two variables (id and friendlyName) that must be default constructed and then initialized separately.  The advantage of the `for` loop is that you can count the total number of parameters passed in case there is an error.

You might consider an implementation that runs in a straight line, without the loop.  Read the `id`.  Then read the `friendlyName`.  If there are leftovers throw an error.  I'm over-simplifying a bit, since you do have to check the tokenizer iterator at each step.  The simplification may be a wash.  And your `for` loop has the advantage that it is easier to extend than a straight-line implementation.

seelabs replied:
Thinking about it now, I think I made a mistake in how this code acts. Everything after the first token should go into the "friendly name". As written, friendly name can not have white space. Not very friendly.

I'll need to keep the loop to do that, but good suggestion otherwise.

scottschurr replied:
It may be worth asking how friendly the name needs to be.  Oftentimes names containing whitespace are more trouble than they are worth.  David or Vinnie may have an opinion on that.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +114,60 @@ class AmendmentTableImpl : public AmendmentTable
+                         "The %1% section in the config file expects %2% "
+                         "items. Found %3%") %
+                     SECTION_AMENDMENTS % numExpectedToks % numToks).str ();
+                throw std::runtime_error (errorMsg);
```
Regardless of whether you change the inner `for` loop to be straight line code, the number of tokens is way less useful to a confused user than the actual text that was parsed (in the `_` variable).  It would probably be better to say something along the lines of "....  Found: '{text in `_`}'.

seelabs replied:
Agree. Will do.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +114,60 @@ class AmendmentTableImpl : public AmendmentTable
+
+                if (numToks == 1)
+                    id.SetHex (curTok);
+                else
```
Can `SetHex()` throw?  What happens if someone passes a non-hex string because they got the parameters reversed on the line?  Will they get a nice error message?

scottschurr replied:
Looking at the old code, it appears that they checked for a bad hex string by testing id for zero after the `SetHex` call.  You probably need to do the same and throw an exception if its zero.  Don't `assert` (what the code used to do).  The error can be caused by user input in the config file.

seelabs replied:
I believe it sets the hash to zero. Will confim and handle this error case, as it is likely to be a common one. As written, addKnown will fail, but the user will not be informed of the error.

seelabs replied:
In case anyone refers to these messages later: SetHex returns a bool that signals success.

seelabs commented:
@nbougalis addKnown was an existing function - I didn't want to change the signature without a good reason (tho I'd rather it didn't return a pointer as well). Will do w.r.t. boost::format. What's rippled's preferred formatter? (I saw boost::format used elsewhere, that's why I picked it). As to return type on line by itself: will do.

nbougalis commented:
@seelabs don't worry about the `boost::format` bits - we can tackle that at another time. Please consider changing `addKnown` to not return a pointer - a simple `bool` seems sufficient to me with the code as is. If we ever need something more, we can add it.

seelabs commented:
@nbougalis Agree about addKnown returning bool. Will do.

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -22,17 +22,30 @@
+
 /** Track the list of "amendments"
 
     An "amendment" is an option that can affect transaction processing
```
Sorry I missed this before!

First, this seems to be the definition of an "amendment".  But you have comments referring to amendments above it in the file.  This definition needs to appear before its first use.

Second, the sentence should be restructured  - it's the most important piece of documentation in this file so it needs to be clear.

First you should say what an amendment _is_: 
`An "amendment" is an option that can affect transaction processing rules.  Amendments are proposed and then adopted or rejected by the network.`

_Then_ you talk about its services and properties! `An Amendment is uniquely identified by its AmendmentID, a 256-bit key.`

Which leads me to my next point - you shouldn't use a "raw" 256-bit ID for Amendments - you should use a _tagged_ 256-bit type.

An example of how to do it is here:  https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/UintTypes.h#L38

The advantage of doing this is that you are prevented from accidentally assigning a `Directory` back and forth with an `Amendment` even though both of them are 256-bit.

By the way, why did we settle on 256-bit rather than 160-bit?  Not that I particularly care - frankly, I think we should have done 256-bit everything :-D - but I'm curious.

seelabs replied:
I made this change to see what it effects and my opinion is it should be a separate ticket. Other places in rippled are assuming an untagged uint256 and making this change seem too unrelated to the original ticket.

I don't think the change is difficult and am happy to make it if you fell strongly we should make it with this ticket.

rec replied:
> Other places in rippled are assuming an untagged uint256

When I first showed up at Ripple, we had *no* tagged uints.  Getting rid of
those was an early assignment of mine - I did all the uint160s but didn't
finish the uint256s because I got diverted to something else, and because
at the time we only had one uint256 type.

It depends really on how much work you think it is.  But I find it much
easier to read an actual type than a generic uint256.

On Tue, Jan 13, 2015 at 6:33 PM, seelabs <notifications@github.com> wrote:

> In src/ripple/app/misc/AmendmentTableImpl.cpp
> <https://github.com/ripple/rippled/pull/795#discussion-diff-22905240>:
>
> >
> >  namespace ripple {
> >
> > +namespace detail
> > +{
> > +/** Amendments that are "compiled into" this app
> > +
> > +    Add amendments to this collection at build time to enable them on this
> > +   server. The first element of the pair is the id, the second element is
> > +   the "friendly name".
> > +*/
> > +std::set<std::pair<std::string, std::string>> PreEnabledAmendments;
> > +}
> > +
> >  /** Track the list of "amendments"
> >
> >      An "amendment" is an option that can affect transaction processing
>
> I made this change to see what it effects and my opinion is it should be a
> separate ticket. Other places in rippled are assuming an untagged uint256
> and making this change seem too unrelated to the original ticket.
>
> I don't think the change is difficult and am happy to make it if you fell
> strongly we should make it with this ticket.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ripple/rippled/pull/795/files#r22905240>.
>

vinniefalco replied:
Ordinarily I would agree but we need to be careful with type proliferation. Types heavily used in business logic (e.g. Account, Book, Currency) are deserving of a tag. Amendments? I'm not sure.

seelabs replied:
I don't think it's a lot of work. I'm arguing that a more focused pull request is better. So this pull request adds amendments (and admittedly did some local cleanup while I was there). If I change to tagged uint256 there non-local cleanups that seem too unrelated to this pull request.

I understand we need to balance extra paperwork (create new ticket, create new pull request w/ reviews). At any rate, not a lot of work. Go head and do it?

vinniefalco replied:
Lets try it without the tag for now, I'd like to see code that is stuck in the pull request process marked Tx Change get modified to use the amendments.

seelabs replied:
@vinniefalco I hear you. I actually agree with @rec that it improves the code, but I definitely fall on the "type all the things!" scale. I programmed in Ada a lone time ago and didn't hate it.

seelabs replied:
@vinniefalco Got it. Will not change. Thanks for the clarification.

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +115,78 @@ class AmendmentTableImpl : public AmendmentTable
-    // addKnown. Set any vetoes or defaults. A pointer to the AmendmentState can be stashed
+    std::set<std::pair<uint256, std::string>> toAdd;
+
+    for (auto const& _ : detail::PreEnabledAmendments)
```
I don't see the advantage of breaking with almost fifty years of tradition in C and C++ of just using a single letter `i`, `a` or the like for loop iterators.

And even though all the compilers I know accept it, it's technically wrong - namespaces starting with an `_` are reserved for the global namespace, which this is not! 

http://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier

scottschurr replied:
While I'm not in favor of using `_` as an identifier, I vaguely remember that this is one of Howard's habits.  I believe a lone underscore squeaks by the rules because it is a leading underscore that is not followed by anything.  @HowardHinnant, am I mis-remembering?  That's not to say that I think a lone `_` is necessarily a *good* name, but it's possibly legal.

HowardHinnant replied:
I don't use `_` in this context but Vinnie has started to.  I have sometimes used a `_` as an identifier when that identifier is not ever referenced.  A typical usage is `lock_guard`:
````
std::lock_guard<std::mutex> _(mut);  // `_` is never referenced and thus has an unimportant name
````
I do not feel strongly about this style and in the past has often caused more trouble than it is worth.  `_` *is* a legal name.  `__` is not.  And `_CaptialLetter` is not legal either.

seelabs replied:
I was trying to follow what I thought as "ripple style". I don't use _ in my own code. I'm actually more comfortable with single-letter loop vars (usually i).

vinniefalco replied:
Use what you're comfortable with then

rec replied:
I really don't believe that `_` is a valid name for development.  From the C++ standard, 

> Each name that begins with an underscore is reserved to the implementation for use as a name in the global namespace.

There is a later rule about names that start `_` and a capital letter but the rule above seems unambiguous.

HowardHinnant replied:
The standard unambiguously reserves `_` for the implementation's use as a name in the global namespace.  But nowhere else.

scottschurr replied:
@rec, that had been my position as well.  My general policy has been to avoid any name that starts with an underscore, since then there's really nothing to worry about.

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +115,78 @@ class AmendmentTableImpl : public AmendmentTable
+                     "PreEnabledAmendments contains an invalid hash (expected "
+                     "a hex number). Value was: %1%") %
+                 _.first).str ();
+            throw std::runtime_error (errorMsg);
```
We're trying to get away from using `boost::format` though this isn't a terrible use compared to some.  Indeed, you might not need the intermediate variable:

    throw std::runtime_error ( "PreEnabledAmendments contains an invalid hash (expected "
                     "a hex number). Value was: " + to_string(a.first));

rec replied:
(As I said below, ignore that comment...)

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +115,78 @@ class AmendmentTableImpl : public AmendmentTable
+            {
+                ++numToks;
+                if (numToks > numExpectedToks)
+                {
```
FYI, you are allowed (but not required) to drop the `{}` in this case.

seelabs replied:
I usually do for a naked continue like this. I left them in because the comment ran over the 80 column limit, and with the comment on two lines I thought the braces were better.

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +115,78 @@ class AmendmentTableImpl : public AmendmentTable
+        // add the amendments from the config file
+        auto const& section = getConfig ().section (SECTION_AMENDMENTS);
+        int const numExpectedToks = 2;
+        for (auto const& _ : section.lines ())
```
Why not `line` instead of `_`?

vinniefalco replied:
If you think about it, an underscore *is* a line.

rec replied:
lol!

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +115,78 @@ class AmendmentTableImpl : public AmendmentTable
+            int numToks = 0;
+            uint256 id;
+            std::string friendlyName;
+            for (auto const& curTok : tokenizer)
```
This logic is tortured.  :-D  The main reason is that you don't know the number of tokens before you start the loop.

What about this sort of thing?

{
    std::vector<std::string> tokens (tokenizer.begin(), tokenizer.end());
    auto size = tokens.size();
    if (size == 0 || size > 2)
    {
        // error return
    }
    else if (size == 1) {
        tokens.push_back (/* compute the human name*/)
    }

    auto const& id = tokens[0];
    auto const& human = tokens[1];

seelabs replied:
Agree, will change. Scott suggested changing this as well, but I wanted the loop in case I decided to support white spaces in the friendly name and the loop didn't seem *that* bad. But two reviews have now objected, I will change (and I agree it makes the code better).

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -100,8 +115,78 @@ class AmendmentTableImpl : public AmendmentTable
+                    if (!id.SetHex (curTok))
+                    {
+                        std::string const errorMsg =
+                            (boost::format (
```
Ah, now I see why you're using `boost::format`!  Forget I made the previous comment about that.  :-)

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -151,28 +236,27 @@ AmendmentTableImpl::get (std::string const& name)
-AmendmentTableImpl::addKnown (const char* amendmentID, const char* friendlyName,
-    bool veto)
+bool
+AmendmentTableImpl::addKnown (uint256 const& hash,
```
You should be passing in `AmendmentName const&` instead of these first two parameters.

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -151,28 +236,27 @@ AmendmentTableImpl::get (std::string const& name)
+bool
+AmendmentTableImpl::addKnown (uint256 const& hash,
+                              std::string const& friendlyName,
+                              bool veto)
```
We strongly recommend not using booleans for these sorts of purposes - because they're hard to audit at the call site.  You could accidentally put `addKnown ({id, name}, true);  // Accept amendment` and your reviewer might just nod eir head.

It's a little verbose but I think you'll come to love using `enum class` entries for this purpose.  In this case you'd have something like:

    enum class Veto {no, yes};
    bool AmendmentTableImpl::addKnown (AmendmentName const& name, Veto veto) //...
    
    // elsewhere:
    table.addKnown (name, Veto::no);

vinniefalco replied:
I'm not a fan of these enum classes just use to give `true` and `false` different names

scottschurr replied:
I'm with Tom on this one (although I haven't sold Vinnie yet).  I've seen cases in rippled code where four bools are being passed through call chains.  It's frightfully easy to mis-order those bools as they go from place to place.  If you make a (very light weight) type (e.g., and enum), then the compiler can tell you whether you've messed up the order.  and you get better naming for free.

If you follow Tom's advice, you might make it an `enum class Veto : unsigned char ...`, since the underlying type of a scoped enum is `int` unless you declare otherwise.

nbougalis replied:
I think that there are cases where an `enum` makes sense. As @scottschurr mentioned, having three or four `bool` variables in a function parameter list is a problem but at the same time I'm not sure I want a proliferation of aliases to `bool`.

Our first instinct should be: why is a `bool` argument needed here? Something smells and can the design be changed to eliminate the argument?

rec commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -125,8 +125,10 @@ class AmendmentTable
 
-    virtual AmendmentState* addKnown (const char* amendmentID,
-        const char* friendlyName, bool veto) = 0;
+    virtual bool addKnown (uint256 const& amendmentID,
```
Comments, comments!  What does this do?  Particularly, what does that inscrutable `bool` return do?

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -151,28 +236,27 @@ AmendmentTableImpl::get (std::string const& name)
     {
         assert (false);
-        return nullptr;
+        return false;
```
As I mentioned elsewhere, it's not clear what the return value means.

Indeed, the semantics of an `assert` should mean that this path is _never_ taken because our program is in a bad state if it is.  So why bother returning anything?

seelabs replied:
I added a comment in the header file to clarify the return value.

returning false to the caller to signal the error seems better than throwing and terminating the app. Even tho this shouldn't happen do we really want to prevent the app from running? I do want that assert hit in a debugger.

Not my code, but I don't think this should change.

rec replied:
We do in fact prevent the app from running - in debug.

But the question is - why do we return the bool?  Does anyone do anything with that?  What we're saying is, "Your program is in a bad state".  Exactly what is the caller supposed to do now that the call failed?

rec replied:
But honestly, this is a quibble.  I don't think this will ever happen, and I don't think the boolean will ever be used and I don't care so much either!

vinniefalco replied:
rippled needs to terminate if a broken invariant is detected, this is critical for ensuring the health of the network. But this needs to be balanced, we also don't want a bug that causes all rippled's to exit at once.

seelabs replied:
I'll change this to throw. Right now it throws in all cases where this assert would be hit anyway (the check just happens before addKnown is called).

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -151,28 +236,27 @@ AmendmentTableImpl::get (std::string const& name)
 
-    AmendmentState* f = getCreate (hash, true);
+    ScopedLockType sl (mLock);
+    AmendmentState* s = getCreate (hash, true);
```
What?!  A naked pointer?  In C++11 and boost, there is literally never a reason to have a naked pointer!

You want nullability?  Then use `boost::optional` on a reference.  It works exactly the same as a naked pointer _except_ that the naked pointer allows of undefined behavior (since a naked pointer's default initialization gives it an undefined value) and `boost::optional` doesn't.

rec replied:
Also, is `s` the right variable name?  It seems that it's sometime like a proxy to an amendment.  Why not just call it `amendment?

nbougalis replied:
I don't care for naked pointers in general, but I think that a naked pointer is fine here - the code here doesn't need to participate in object lifetime decisions so doesn't need `unique_ptr` or `shared_ptr` and the lifetime of the object being pointed to is guaranteed by external code. Additionally, the code in question is a pure implementation detail and the pointers don't escape. 

I guess this could, in theory, be a reference (and throwing from `getCreate` in the unlikely event that creation failed) but I wouldn't worry too much about this.

rec commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -151,28 +236,27 @@ AmendmentTableImpl::get (std::string const& name)
-    if (friendlyName != nullptr)
-        f->setFriendlyName (friendlyName);
+    if (!friendlyName.empty ())
+        s->setFriendlyName (friendlyName);
```
Note that you're using `s` without checking that it's non-null.  Or if you know it's non-null, then you should use a reference instead.

(You might have the convention that mutable objects are always referred to by pointers, which was my pre-ripple custom, but we make free use of references here.  I confess that it's quite useful...)

seelabs commented:
I think I addressed all the feedback. The only two I did add to the code were changing the amendment id from a uint256 to a tagged base_uint type and a  comment about asserting. I left comments above as to why. If you think I missed your feedback point me to the comment and I'll gladly look again.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,24 +99,86 @@ class AmendmentTableImpl : public AmendmentTable
 {
-    // For each amendment this version supports, construct the AmendmentState object by calling
-    // addKnown. Set any vetoes or defaults. A pointer to the AmendmentState can be stashed
+    detail::PreEnabledAmendmentsCollection toAdd (detail::preEnabledAmendments);
```
I like what I've seen so far.  Here's the smallest of nits.  You might consider moving the `toAdd` local lower in the method -- below the following for loop.  That way if there are errors in `preEnabledAmendments` we won't even bother constructing `toAdd`.  It also reduces `toAdd`'s scope slightly.

seelabs replied:
Gah! When I was refactoring I introduced a bug. Fixing now (I need to add "a" to the collection in the loop). Thanks for pointing this out.

seelabs replied:
(Thought there was a bug for a second. It's OK). Will do.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,24 +99,86 @@ class AmendmentTableImpl : public AmendmentTable
-AmendmentState*
-AmendmentTableImpl::getCreate (uint256 const& amendmentHash, bool create)
+AmendmentState&
+AmendmentTableImpl::getCreate (uint256 const& amendmentHash)
```
Well done, getting rid of the `bool` parameter.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -192,7 +267,7 @@ bool
 {
     ScopedLockType sl (mLock);
-    AmendmentState* s = getCreate (amendment, false);
+    AmendmentState* s = getExisting (amendment);
```
Using the separate `getCreate` and `getExisting` methods makes the code much easier to read.  Nicely done.

nbougalis replied:
I agree - much cleaner. For future reference, please prefer `auto s = getExisting (amendment);`

scottschurr replied:
FWIW, Nik and I agree to disagree on whether `auto` is preferable.  (Yes, I'm swimming against the tide on this one.)  I like being able to look at the source code and know that, in this particular case, I have a raw pointer in my hands.  That information is hidden by `auto`.  And my IDE doesn't tell me.  Nik has more seniority, though ;-).

I'm not always against `auto`.  It's good in generic situations or if the name of a type is just too ugly to easily read (many iterator names).  Those don't match this case.

scottschurr commented:
I'm wondering if anyone has done a lock audit on `AmendmentTableImpl`?.  I have the sense that the `get()` method should probably hold a lock since it might be searching `m_amendmentMap` while the map is mutating.

nbougalis commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -22,17 +22,18 @@
+class AmendmentTableImpl final : public AmendmentTable
 {
 protected:
     typedef hash_map<uint256, AmendmentState> amendmentMap_t;
```
The `amendmentIt_t` can now be removed.

nbougalis commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -50,7 +51,8 @@ class AmendmentTableImpl : public AmendmentTable
 
-    AmendmentState* getCreate (uint256 const& amendment, bool create);
+    AmendmentState& getCreate (uint256 const& amendment);
+    AmendmentState* getExisting (uint256 const& amendment);
```
Why does `getCreate` return a reference and `getExisting` a pointer? That seems almost reversed in a way, but I'd argue they should both return the same type.

nbougalis replied:
I'm not sure I like the asymmetry of return types here, to be honest.

seelabs replied:
The problem is what "getExisting" should do if the element it's asking for isn't in the table. I don't to want getExisting to thow - it goes against how the rest of rippled uses exceptions. Not sure how to address this one.

vinniefalco replied:
Its good as written. The different return types are no big deal. The alternative is to return an iterator and provide begin/end functions. Which we should consider if significant new code calls the public interfaces of this class.

nbougalis commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -53,19 +95,25 @@ class AmendmentSet
     AmendmentState ()
-        : mVetoed (false), mEnabled (false), mSupported (false), mDefault (false),
-          m_firstMajority (0), m_lastMajority (0)
+        : mVetoed (false)
```
If all these variables were "inline initialized" then the constructor is reduced to: `AmendmentState () = default`

nbougalis commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -53,19 +95,25 @@ class AmendmentSet
 {
 public:
-    bool mVetoed;   // We don't want this amendment enabled
+    bool mVetoed;  // We don't want this amendment enabled
```
Inline class intialization for these would be nice: `bool mVetoed = false;`

seelabs commented:
@nbougalis I added all the changes except changing the asymmetry of the return types. Left a note above as to why.

scottschurr commented:
@seelabs: I'll give it a thumbs up if you write a comment why `AmendmentTableImpl::get` doesn't need a lock or you put the lock in place.  Sometimes David says that locks aren't needed in certain places, but we should always document why that's the case.  I doubt this is one of David's special cases, however.

seelabs commented:
@scottschurr get is a public function that iterates through the table. You're right, it needs a lock. Done. I also looked at other code that might check out a lock and call get, and there wasn't any.

scottschurr commented:
:+1:
Looks great!

vinniefalco commented:
Does the unit test instantiate the class and exercise its public members?

seelabs commented:
Writing unit test now. Needed to change the interface a little so I can pass in a config.

rec commented:
:+1: conditional on a unit test.  Good work on catching the lock and on eliminating that `bool` parameter altogether!  One less thing to have to know about...

rec commented:
Oh, one more comment.

I really believe that there is _no_ use case for naked pointers any more - at least in rippled's flavor of C++11.

Often in the rippled codebase, when we return a naked pointer it's because we're _creating_ that pointer - so the caller owns the pointer.  We're converting these as we see them to `std::unique_ptr` - that's a no-brainer.

Now, look at `AmendmentState* getExisting (uint256 const& amendment);` - who owns the returned `AmendmentState*`?  is it nullable?

Now I can read the answers by reading the code but returning `boost::optional<AmendmentState&> getExisting (uint256 const&);` makes it unnecessary to even ask the question in the first place.

Just my $0.02 on this issue!

vinniefalco commented:
Pointers are part of the language, and since we require C++11 , the assumption that naked pointers do not imply ownership is reasonable. There are use-cases for naked pointers, this looks like one of them.

rec commented:
Then add a comment saying that the caller does _not_ own the returned pointer, and that nullptr indicates that an Amendment wasn't found, please!

rec commented:
Or really, let's just leave it alone.  :-)

seelabs commented:
I checked in a unit test for addInitial, but I have not checked in the supporting files (to come, want to review locally first). Does run and tests do pass.

nbougalis commented:
@rec on the topic of naked pointers, my thoughts are that if you need to participate in object lifetime management, then use `std::unique_ptr` and `std::shared_ptr` as appropriate. Otherwise, passing a pointer is acceptable although I'd prefer a reference if the thing pointed to is guaranteed to never be null.

rec commented:
that last change seemed to cause a legitimate compilation error in clang... :-(

seelabs commented:
I added unit tests for some of the public interface to AmendmentTable. The unit test and the way I'm doing dependency injection into the class needs to be reviewed. After that, I'd like to mark this as ready to merge. 

There are some public functions of Amendment table that do not yet have unit tests, however, I would like to make those a separate ticket and submit them with another PR.

I will squash all the small commits into one commit when we mark this as ready to merge.

scottschurr commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -154,10 +211,40 @@ class AmendmentTable
-make_AmendmentTable (std::chrono::seconds majorityTime, int majorityFraction,
-    beast::Journal journal);
+/**
+   AmendmentTableInjections is used to insert moc objects into the amendment
```
Hi Scott, no objection here.  But just for my own curiosity, I've always seen it spelled "mock", like mock turtle.  Is "moc" also a common spelling?

scottschurr commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -154,10 +211,40 @@ class AmendmentTable
+    virtual void setMajorityTimesFromDBToState (
+        AmendmentState& toUpdate,
+        uint256 const& amendmentHash) const = 0;
+    /** For eash hash, get the first and last majority from the corresponding
```
"eash" -> "each"?

scottschurr commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -154,10 +211,40 @@ class AmendmentTable
+// Use for regular system
+std::unique_ptr<AmendmentTableInjections> make_AmendmentTableInjections ();
+// Use for unit testing
+std::unique_ptr<AmendmentTableInjections> make_MOCAmendmentTableInjections ();
```
Again, just a nit, but "MOC" -> "MOCK"?

seelabs commented:
"Mock" is correct. Will change ("MOC" ate into my brain from QT programming).

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,51 +101,110 @@ class AmendmentTableImpl : public AmendmentTable
+                     "preEnabledAmendments contains an invalid hash (expected "
+                     "a hex number). Value was: %1%") %
+                 a.hexString ()).str ();
+            throw std::runtime_error (errorMsg);
```
Thoughts for the future.  If there's an error in `preEnabledAmendments` that's a programming error, right?  So you might consider an additional `assert` here so we get stopped dead in debug mode.  You have to leave the `throw`, because the `assert` will get compiled out in release mode.

Also, for my own education, is `boost::format` an improvement over what is, basically, string concatenation?

seelabs replied:
Some of the other error messages were more involved than this one. I used boost::format for those. This one could easily be be replaced with string concat, but I like that it follows the same pattern as the other error handlers.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,51 +101,110 @@ class AmendmentTableImpl : public AmendmentTable
+    detail::PreEnabledAmendmentsCollection toAdd (detail::preEnabledAmendments);
 
+    {
+        // add the amendments from the config file
```
Less than a nit: My friend, Mr. Swirly, has berated me for non-programming text (e.g., comments) that doesn't have an initial cap and a closing period.  I think he has a point about improved readability.  I've been monitoring my own comments so I'm less likely to get hammered if he reviews my code.  You might consider the same habit.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,51 +101,110 @@ class AmendmentTableImpl : public AmendmentTable
+            }
 
-            if (db->executeSQL (query) && db->startIterRows ())
+            toAdd.emplace_back (std::move (tokens[0]), std::move (tokens[1]));
```
Nice use of `emplace`.  I need to develop that habit.

I also find this code much easier to read than the original nested `for`.  Good change.

scottschurr commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -378,22 +440,9 @@ AmendmentTableImpl::reportValidations (const AmendmentSet& set)
-        }
-        db->executeSQL ("END TRANSACTION;");
-        changedAmendments.clear();
+        m_injections->setMajorityTimesFromStateToDB (changedAmendments,
```
Implementation in mock looks like it matches the old implementation.

scottschurr commented:
:+1:
Looks great.  Nice job with the unit tests too.

vinniefalco commented:
There's a merge commit at the top of the branch

seelabs commented:
Thanks Scott. I'll clean up all the small git commits into one commit after I get the other thumbs up.

vinniefalco commented:
I'm not really feeling these names like `AmendmentTableInjections`

seelabs commented:
@vinniefalco I needed to merge after I pushed a minor cleanup that it couldn't automatically merge (probably due to the way I brought in develop into the repo). I'll figure out how to clean up the mess and make a clean commit after I get the other thumbs up.

seelabs commented:
@vinniefalco I could split the class into two: one dealing with the walletDB interface and one dealing with getValidations. Then naming would be easier: `AmendmentTableWalletDB` and `AmendmentTableValidations` or somesuch.

I'd prefer to keep the single class tho. Injection isn't great, I agree, but tough to come up with a name for a class like this.

scottschurr commented:
I needed to create a similar mock in `TransactionSign.h`.  That mock provides even more scattered functionality than this one does.  We ended up naming the mock `TxnSignApiFacade` (I believe a suggestion from Vinnie).  I'm pretty satisfied with that name.  You might consider something similar.

vinniefalco commented:
Why does the facade need to be wrapped in `shared_ptr`?

seelabs commented:
@vinniefalco You mean a unique_ptr? Or was that comment directed at ScottS?

I'm not sure what you're suggesting.

seelabs commented:
@vinniefalco I got rid of the unique_ptr. The facade is not a regular object in the class. I'm guessing that's what you meant with your comment about shared_ptr.

vinniefalco commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -154,10 +211,26 @@ class AmendmentTable
-std::unique_ptr<AmendmentTable>
-make_AmendmentTable (std::chrono::seconds majorityTime, int majorityFraction,
-    beast::Journal journal);
+namespace AmendmentTableDetail
```
`namespace detail` is good enough, no need to invent a new identifier for every detail namespace.

vinniefalco commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -154,10 +211,26 @@ class AmendmentTable
+enum class AppApiFacade
+{
+    useApp,
+    useMock
```
There's no need to proliferate identifiers used in such a narrow scope, when a bool suffices. Although if its in a detail namespace I am less concerned. Its still overly verbose. But looking at the bigger picture, a bool is inferior to an abstract interface. The implementation should not be aware of whether or not it is being passed a mock interface.

vinniefalco commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -49,26 +49,30 @@ class AmendmentTableImpl : public AmendmentTable
         , m_firstReport (0)
         , m_lastReport (0)
-        , m_journal (journal)
+        , m_journal (std::move (journal))
```
journals are lightweight and copyable, no need to move them or pass them by reference.

vinniefalco commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,51 +101,114 @@ class AmendmentTableImpl : public AmendmentTable
 {
-    // For each amendment this version supports, construct the AmendmentState object by calling
-    // addKnown. Set any vetoes or defaults. A pointer to the AmendmentState can be stashed
+using PreEnabledAmendmentsCollection = std::vector<AmendmentName>;
```
Is "PreEnabledAmendmentsCollection" really preferred over "std::vector<AmendmentName>;" ??

vinniefalco commented on src/ripple/app/misc/tests/AmendmentTable.test.cpp:
```diff
@@ -0,0 +1,391 @@
+             {TablePopulationAlgo::addInitial, TablePopulationAlgo::addKnown})
+        {
+            {
+                // test that the amendmens we add are enabled and amendments we
```
spelling

vinniefalco commented on src/ripple/core/ConfigSections.h:
```diff
@@ -30,8 +30,8 @@ struct ConfigSection
 
-// VFALCO TODO Rename and replace these macros with variables.
 #define SECTION_ACCOUNT_PROBE_MAX       "account_probe_max"
+#define SECTION_AMENDMENTS              "amendments"
```
We need to be removing macros from this file not adding them!

vinniefalco commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,51 +101,114 @@ class AmendmentTableImpl : public AmendmentTable
 
-void
-AmendmentTableImpl::addInitial ()
+namespace detail
```
detail namespaces are for header files not for .cpp files. The contents of .cpp files are already implicitly private, identifiers do not need additional markings. A detail namespace is used when it is necessary for public interface declarations in headers to have visibility on private identifiers. This said, there's no harm using a detail namespace in a .cpp file but it adds no value.

scottschurr replied:
@vinniefalco I disagree with your statement that detail namespaces are for header files only.  In a previous life I made extensive use of anonymous namespaces in cpp files.  Our unity builds make anonymous namespaces ineffective.  In our code base I have replaced using anonymous namespaces with `detail` namespaces in cpp files.  In effect, I'm hand-building a pseudo-anonymous namespace.  Can you suggest an alternative?

vinniefalco commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -97,51 +101,114 @@ class AmendmentTableImpl : public AmendmentTable
-    // addKnown. Set any vetoes or defaults. A pointer to the AmendmentState can be stashed
+using PreEnabledAmendmentsCollection = std::vector<AmendmentName>;
+
+/** preEnabledAmendments is a static collection of amendments are are enabled at build time.
```
Grammar

vinniefalco commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -569,19 +643,116 @@ AmendmentTableImpl::getJson (uint256 const& amendmentID)
+    std::chrono::seconds majorityTime,
+    int majorityFraction,
+    beast::Journal journal,
+    AmendmentTableDetail::AppApiFacade appApiFacade)
```
I do not understand the need for all of this class enum rigamarole. First of all, callers would need to use declarations in a detail namespace in order to call this function, defeating the purpose. Second, the name of the enum matches the formal template argument name in the implementation class, a point of potential confusion. Third, factory functions like this should not contain business logic. Consider this declaration instead:
```
template <class Rep, class Period,
    class AppType = AmendmentTableDetail::AppApiFacadeImpl>>
std::unique_ptr<AmendmentTable>
make_AmendmentTable (std::chrono::duration<Rep, Period> const& majorityTime,
    int majorityFraction, beast::Journal journal);
```

Normal invocations use `AppApiFacadeImpl` while test code may use this alternate invocation:
```
...
make_AmendmentTable<AmendmentTableDetail::AppApiFacadeMock>(
    majorityTime, journal); 
...
```

seelabs replied:
Will do.

vinniefalco replied:
We don't even need my modified version. `make_AmendmentTable` is a public interface and should always create a version of the implementation that uses the "real" Application object, while the unit test can merely include the private implementation class declaration header and call `make_shared` directly, using the mock type.

vinniefalco commented on src/ripple/app/misc/tests/AmendmentTable.test.cpp:
```diff
@@ -0,0 +1,391 @@
+    std::vector<AmendmentName> populateTable (
+        AmendmentTable& table,
+        StringPairVec const& amendmentPairs,
+        TablePopulationAlgo populationAlgo = TablePopulationAlgo::addKnown)
```
All these long names and type proliferations are giving me a headache

vinniefalco commented on src/ripple/app/misc/tests/AmendmentTable.test.cpp:
```diff
@@ -0,0 +1,391 @@
+                    table.addKnown (i);
+                }
+                break;
+            case TablePopulationAlgo::addInitial:
```
grrr...we added all these identifiers for just 2 ways to fill the table?

vinniefalco commented on src/ripple/app/misc/tests/AmendmentTable.test.cpp:
```diff
@@ -0,0 +1,391 @@
+        return make_AmendmentTable (
+            weeks (2),
+            majorityFraction,
+            journal,
```
The journal could be a default argument in `make_AmendmentTable`

seelabs replied:
If we got rid of the final bool parameter - as in your template design - that makes sense. Since I'm keeping the final parameter (for now) we don't gain anything by making journal a default parameter.

vinniefalco commented on src/ripple/app/main/Application.cpp:
```diff
@@ -652,7 +654,8 @@ class ApplicationImp
 
-        m_amendmentTable->addInitial();
+        m_amendmentTable->addInitial (
+            getConfig ().section (SECTION_AMENDMENTS));
```
Why have `SECTION_AMENDMENTS` if its only used in one place?

seelabs commented:
@vinniefalco I did all the minor cleanups you suggested. For now, I'd rather keep the implementation in  a cpp file instead of moving it to a header (as suggested elsewhere). That means keeping the `make_AmendmentTable` with a bool parameter. I might go back and refactor this later - especially as we add more testing. I'm anxious to close this ticket.

rec commented:
Me too, where are we?  

I was pretty pro-this the last time.  Vinnie, are all your objections satisfied so I can give it a last once-over?

vinniefalco commented:
@rec Its about to be merged so you better be quick, I'm ok with it. Although we will almost certainly need to revisit it later.

rec commented:
Great, that's good news.  No further comments!

seelabs commented:
@vinniefalco Want me to squash all those small commits now?

vinniefalco commented:
@rec, @seelabs This has already been squashed down and is about to be merged as *0.28.0-b1*
https://github.com/ripple/rippled/pull/829

@seelabs did you change it since yesterday morning?

seelabs commented:
@vinniefalco No, I haven't touched it.

nbougalis commented:
Merged as 44450bf6448c56ca3f218ba726a9b4e33d422e51
